tidy3d/components/simulation.py

```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
             elif medium.allow_gain:
                 return True
         return False
+
+    @staticmethod
+    def _sel_data_inside_bounds(data: SpatialDataArray, bounds: Bound) -> Tuple[bool, SpatialDataArray]:
+
+        if data is None:
+            return True, None
+
+        is_full_cover = True
+
+        inds_list = []
+
+        for coord, smin, smax in zip(
+            data.coords.values(),
+            bounds[0],
+            bounds[1]
+        ):
+
+            length = len(coord)
+
+            # if data does not cover structure at all take the closest index
+            if smax < coord[0]:  # structure is completely on the left side
+
+                # take 2 if possible, so that linear iterpolation is possible
+                comp_inds = np.arange(0, max(2, length))
+                is_full_cover = False
+
+            elif smin > coord[-1]:  # structure is completely on the right side
+
+                # take 2 if possible, so that linear iterpolation is possible
+                comp_inds = np.arange(min(0, length - 2), length)
+                is_full_cover = False
+
+            else:
+                if smin < coord[0]:
+                    is_full_cover = False
+                    ind_min = 0
+                else:
+                    ind_min = max(0, (coord >= smin).argmax().data - 1)
+
+                if smax > coord[-1]:
+                    is_full_cover = False
+                    ind_max = length - 1
+                else:
+                    ind_max = (coord >= smax).argmax().data
+
+                comp_inds = np.arange(ind_min, ind_max + 1)
+
+            inds_list.append(comp_inds)
+
+        return is_full_cover, data.isel(x=inds_list[0], y=inds_list[1], z=inds_list[2])
+
+    def apply_perturbations(
+        self,
+        temperature: SpatialDataArray = None,
+        electron_density: SpatialDataArray = None,
+        hole_density: SpatialDataArray = None,
+    ) -> Simulation:
+        """Apply heat and/or charge data to perturbation mediums contained in the Simulation
```

Comment:
let's make sure the users know that this returns a copy, so they dont do
```py
sim.apply_perturbations(...)
```
and not realize why things didn't change.  maybe instead of "Apply heat and/or charge data to ..." we should word it as "Return a copy of the simulation with heat and/or charge data applied to ...

Reply:
Agree. I also renamed it to `.perturbed_mediums_copy()` by analogy to `.updated_copy()`